services/pull/merge.go

```diff
@@ -30,8 +31,93 @@ import (
 	"code.gitea.io/gitea/modules/timeutil"
 	asymkey_service "code.gitea.io/gitea/services/asymkey"
 	issue_service "code.gitea.io/gitea/services/issue"
+
+	"gitea.com/lunny/size"
+	"github.com/unknwon/com"
 )
 
+// GetDefaultMergeMessage returns default message used when merging pull request
+func GetDefaultMergeMessage(baseGitRepo *git.Repository, pr *models.PullRequest, mergeStyle repo_model.MergeStyle) string {
+	if err := pr.LoadHeadRepo(); err != nil {
+		log.Error("LoadHeadRepo[%d]: %v", pr.HeadRepoID, err)
+		return ""
+	}
+	if err := pr.LoadIssue(); err != nil {
+		log.Error("Cannot load issue %d for PR id %d: Error: %v", pr.IssueID, pr.ID, err)
+		return ""
+	}
+	if err := pr.LoadBaseRepo(); err != nil {
+		log.Error("LoadBaseRepo: %v", err)
+		return ""
+	}
+
+	isExternalTracker := pr.BaseRepo.UnitEnabled(unit.TypeExternalTracker)
+	issueReference := "#"
+	if isExternalTracker {
+		issueReference = "!"
+	}
+
+	if mergeStyle != "" {
+		templateFilepath := fmt.Sprintf(".gitea/default_merge_message/%s_TEMPLATE.md", strings.ToUpper(string(mergeStyle)))
+		commit, err := baseGitRepo.GetBranchCommit(pr.BaseRepo.DefaultBranch)
+		if err != nil {
+			log.Error("GetBranchCommit: %v", err)
+			return ""
+		}
+		templateContent, err := commit.GetFileContent(templateFilepath, int(10*size.K))
+		if err != nil {
+			log.Error("GetFileContent: %v", err)
+			return ""
+		}
+
+		vars := map[string]string{
+			"BaseRepoOwnerName":      pr.BaseRepo.OwnerName,
+			"BaseRepoName":           pr.BaseRepo.Name,
+			"BaseBranch":             pr.BaseBranch,
+			"HeadRepoOwnerName":      pr.HeadRepo.OwnerName,
+			"HeadRepoName":           pr.HeadRepo.Name,
+			"HeadBranch":             pr.HeadBranch,
+			"PullRequestTitle":       pr.Issue.Title,
+			"PullRequestDescription": pr.Issue.Content,
+			"PullRequestPosterName":  pr.Issue.Poster.Name,
+			"PullRequestIndex":       strconv.FormatInt(pr.Index, 10),
+			"PullRequestReference":   fmt.Sprintf("%s%d", issueReference, pr.Index),
+			"IssueReferenceChar":     issueReference,
```

Comment:
Do we still need this `IssueReferenceChar`?

And since some var names are changed, the document could be updated too.